### DISCUSSION TITLE
2D Two-Phase Drainage Simulation

### DISCUSSION THREAD

**Discussion by PNMExplorer** [PMEAL/OpenPNM#2825] [Q&A] [answered]
Hello OpenPNM Community,

I'm currently working on a project involving a 2D TIFF image of a pore network and the implementation of a two-phase drainage simulation in OpenPNM. In this simulation, I aim to incorporate contact angle modeling, define viscosities for both air and water, and apply a pressure difference. I have explored the OpenPNM example regarding drainage, but I noticed that it primarily focuses on creating random shapes. However, my specific challenge is to adapt this to work with a pore network extracted from an image.

![unais_pm1-1](https://github.com/PMEAL/OpenPNM/assets/145595228/247e9013-febf-4246-9bee-46408b8cf6b2)

I've made several attempts to extract the network from the 2D single image, but I've encountered persistent errors. I'm seeking guidance on how to approach this task in OpenPNM, and I'm also curious if anyone in the community has a similar code or script they'd be willing to share as a reference. Your expertise, insights, and the possibility of code sharing would be immensely helpful in resolving these issues and progressing with this simulation.

Thank you all for your assistance!

**Comment by jgostick** [comment]
You'll want to look at porespy, which is OpenPNM's sibling package that contains all the image analysis tools including network extraction.  Specifically [this example](https://porespy.org/examples/networks/tutorials/snow_basic.html) and [this one](https://porespy.org/examples/networks/reference/snow2.html)

**Comment by likeduck** [comment, marked as answer]
Hi this is [a manuscript](https://github.com/likeduck/SimulationExample-OpenPNM/blob/main/PoreSpy/SNOW.py), but the effect is not ideal. It may be because your image throat is too thin.
![Figure_1](https://github.com/PMEAL/OpenPNM/assets/117707506/352b5133-8442-438d-806e-20952591273d)

**Reply by PNMExplorer** [reply]
Thank you @likeduck for your reply. I also would like to ask you another question if possible. OpenPNM is dynamic right? but why is it not possible to change the flowrate or pressure difference. No matter what I change the value of flowrate to, it doesn't affect the result. Also, I would like to know if you have run any 3D two-phase simulation before like for example have you extracted a pore network from a rock sample from 'Digitalrockportal' and run a drainage or imbibition simulation with time step. Like how it invades from one end to other. If you did, can you please share an example code with me. Also, I want to know how long it takes from to finish the simulation when using time step. One last question, is it possible to change the viscosity ratio when calculating permeability in openPNM? I tried to change the viscosity ratio, but it doesn't change.

I would really appreciate if you could help me with my doubts. Thank you

**Reply by jgostick** [reply]
Invasion percolation and drainage in OpenPNM are not dynamic.  All of our algorithms are 'quasi-static' meaning fluid configuration is entirely controlled by capillary effects.  We have never had the need in our own research to do an invasion simulation where the flow rate was high enough for viscous effects to matter, so never had a chance/need to develop such an algorithm.  Sorry.

**Reply by PNMExplorer** [reply]
Ok. Thank you @jgostick for the clarification.
